### DISCUSSION TITLE
Terminal Gui Designer

### DISCUSSION THREAD

**Discussion by tznind** [tui-cs/Terminal.Gui#1641] [Show and tell]
2022-06-18
Update adds support for multi selection.  Although currently the only operations that support multi select are delete and moving views with keyboard (Shift+Cursor).

![multiselect](https://user-images.githubusercontent.com/31306100/174438149-069306e2-739b-4ff3-b363-260374aa6215.gif)

2022-05-08
Latest progress includes adding support for designing MenuBar instances.

![menus](https://user-images.githubusercontent.com/31306100/167308558-9023da0b-de3b-4d34-b243-d12010e8b88a.gif)

2022-04-22
Made quite a lot more progress on this.  Still a lot of stuff to implement but thought I'd update since its come quite a way since this first post.  

![demo2](https://user-images.githubusercontent.com/31306100/164798580-13456eaa-6154-4fe6-8b31-27413fecb5f1.gif)

--------------------------------------------------------------------------------

I've pushed a version 1.0.5 to nuget which you can install as a CLI tool with:

```
dotnet tool install --global TerminalGuiDesigner
```

--------------------------------------------------------------------------------

I've started exploring writing a tool like the Windows Forms Designer for gui.cs.  I've gotten some basics done using CodeDom for C# code generation (saving) and Roslyn for loading.  I had a look and didn't see any existing designers but let me know if there are any already.

There's a lot to do but thought I'd show what I had done so far.  The repo is https://github.com/tznind/TerminalGuiDesigner .

![designer](https://user-images.githubusercontent.com/31306100/161325121-c6c03350-5d37-4830-b756-58daf79c972f.gif)

**Comment by BDisp** [comment]
That's very cool @tznind. Someday it will be very easy creating a `Terminal.Gui` app.

**Comment by tznind** [comment]
Added support for dragging into different container views.  Now I probably need a way to do the same with the keyboard, I'd like to maintain usability with keyboard only if possible.

![dragging-into](https://user-images.githubusercontent.com/31306100/169541098-8354f00e-39e6-4d83-bc80-48bd6c6c5f26.gif)
_Dragging into different containers then hammering undo/redo_

**Comment by tig** [comment]
This is beyond amazeballs. Can't wait to get real Internet connectivity so I can try it!

**Reply by tznind** [reply]
Thanks! I'd be really grateful to hear your feedback on usability and any issues.  

I think theres some more work to make sure its fully bulletproof but once its at that stage it would be great if we could add a link to it from the Terminal.Gui readme to help with discoverability if thats appropriate.
